docs/upload-strategies/aws-s3.mdx

```diff
@@ -0,0 +1,403 @@
+
+This is passed through to [XHRUpload](/docs/xhrupload#timeout-30-1000); see its documentation page for details.
+Set to `0` to disable this check.
+
+#### `limit`
+
+Limit the amount of uploads going on at the same time (`number`, default: `5`).
+
+Setting this to `0` means no limit on concurrent uploads, but we recommend a value between `5` and `20`.
+
+#### `getResponseData(responseText, response)`
+
+:::note
+This is an advanced option intended for use with _almost_ S3-compatible storage solutions.
+:::
+
+Customize response handling once an upload is completed.
+This passes the function through to @uppy/xhr-upload,
+see its [documentation](https://uppy.io/docs/xhr-upload/#getResponseData-responseText-response) for API details.
+
+This option is useful when uploading to an S3-like service that doesn’t reply with an XML document,
+but with something else such as JSON.
+
+#### `locale: {}`
+
+```js
+export default  {
+  strings: {
+    timedOut: 'Upload stalled for %{seconds} seconds, aborting.',
+  },
+}
+```
+
+## Frequently Asked Questions
+
+### How can I generate a presigned URL server-side?
+
+The `getUploadParameters` function can return a Promise, so upload parameters can be prepared server-side.
+That way, no private keys to the S3 bucket need to be shared on the client.
+For example, there could be a PHP server endpoint that prepares a presigned URL for a file:
+
+```js
+uppy.use(AwsS3, {
+  getUploadParameters (file) {
+    // Send a request to our PHP signing endpoint.
+    return fetch('/s3-sign.php', {
+      method: 'post',
+      // Send and receive JSON.
+      headers: {
+        accept: 'application/json',
+        'content-type': 'application/json',
+      },
+      body: JSON.stringify({
+        filename: file.name,
+        contentType: file.type,
+      }),
+    }).then((response) => {
+      // Parse the JSON response.
+      return response.json()
+    }).then((data) => {
+      // Return an object in the correct shape.
+      return {
+        method: data.method,
+        url: data.url,
+        fields: data.fields,
+        // Provide content type header required by S3
+        headers: {
+          'Content-Type': file.type,
+        },
+      }
+    })
+  },
+})
+```
+
+See the [aws-presigned-url example in the uppy repository](https://github.com/transloadit/uppy/tree/main/examples/aws-presigned-url) for a small example that implements both the server-side and the client-side.
+
+### How can I retrieve the presigned parameters of the uploaded file?
+
+Once the file is uploaded, it’s possible to retrieve the parameters that were
+generated in `getUploadParameters(file)` via the `file.meta` field:
+
+```js
+uppy.on('upload-success', (file, data) => {
+  const s3Key = file.meta['key'] // the S3 object key of the uploaded file
+})
+```
+
+### What makes a provider S3 compatible?
+
+...
```

Comment:
@mifi do you know a good answer to this perhaps?

Reply:
I think this has been answered in the existing docs:

https://uppy.io/docs/aws-s3/#S3-alternatives

although I think we should add a note saying something like we don’t officially support and/or test against any of these s3-providers, so it’s not a first-class support, but it should work. (I think I remember issues with s3-like providers before, where we decided not to implement special code to make it work with those)

Reply:
That doesn't answer this question though, I mean how do you know if random provider X is S3 compatible? What does it take to be S3 compatible?

Regarding the note, that's already in "when should I use this?" section

Reply:
I don't really have a good answer to this question. It would require a lot of testing to see if all functionality that Uppy needs is supported by each s3-like provider, also at scale. Or we can just say that we don't provide a list of supported s3-like providers, but we treat s3 with duck-typing in mind. "If it talks like s3 and it uploads like s3, then it's probably s3."

Reply:
```suggestion
```